controllers/openstackdataplanedeployment_controller.go

```diff
@@ -0,0 +1,265 @@
+
+	// Deploy each nodeSet
+	// The loop starts and checks NodeSet deployments sequentially. However, after they
+	// are started, they are running in parallel, since the loop does not wait
+	// for the first started NodeSet to finish before starting the next.
+	shouldRequeue := false
+	haveError := false
+	for _, nodeSet := range nodeSets.Items {
+
+		logger.Info(fmt.Sprintf("Deploying NodeSet: %s", nodeSet.Name))
+		logger.Info("Set Status.Deployed to false", "instance", instance)
+		instance.Status.Deployed = false
+		logger.Info("Set DeploymentReadyCondition false", "instance", instance)
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.DeploymentReadyCondition, condition.RequestedReason,
+			condition.SeverityInfo, condition.DeploymentReadyRunningMessage))
+		ansibleEESpec := nodeSet.GetAnsibleEESpec()
+		ansibleEESpec.AnsibleTags = instance.Spec.AnsibleTags
+		ansibleEESpec.AnsibleSkipTags = instance.Spec.AnsibleSkipTags
+		ansibleEESpec.AnsibleLimit = instance.Spec.AnsibleLimit
+
+		nodeSetSecretInv := fmt.Sprintf("dataplanenodeset-%s", nodeSet.Name)
+
+		if nodeSet.Status.DNSClusterAddresses != nil && nodeSet.Status.CtlplaneSearchDomain != "" {
+			ansibleEESpec.DNSConfig = &corev1.PodDNSConfig{
+				Nameservers: nodeSet.Status.DNSClusterAddresses,
+				Searches:    []string{nodeSet.Status.CtlplaneSearchDomain},
+			}
+		}
+
+		deployResult, err := deployment.Deploy(
+			ctx, helper, &nodeSet, instance,
+			nodeSetSecretInv, &instance.Status, ansibleEESpec,
+			nodeSet.Spec.Services)
+
+		if err != nil {
+			util.LogErrorForObject(helper, err, fmt.Sprintf("OpenStackDeployment error for NodeSet %s", nodeSet.Name), instance)
+			haveError = true
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				condition.ReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				dataplanev1.DataPlaneNodeSetErrorMessage,
+				err.Error()))
+		}
+
+		if deployResult != nil {
+			shouldRequeue = true
+		} else {
+			logger.Info("OpenStackDeployment succeeded for NodeSet", "NodeSet", nodeSet.Name)
+			logger.Info("Set NodeSetDeploymentReadyCondition true", "nodeSet", nodeSet.Name)
+			instance.Status.Conditions.Set(
+				condition.TrueCondition(
+					condition.Type(fmt.Sprintf(dataplanev1.NodeSetDeploymentReadyCondition, nodeSet.Name)),
+					condition.DeploymentReadyMessage))
+
+			logger.Info("Set NodeSet DeploymentReadyCondition true", "nodeSet", nodeSet.Name)
+			_, err = controllerutil.CreateOrPatch(ctx, helper.GetClient(), &nodeSet, func() error {
+				nodeSet.Status.Conditions.MarkTrue(condition.DeploymentReadyCondition, condition.DeploymentReadyMessage)
+				return nil
+			})
+			if err != nil {
+				util.LogErrorForObject(helper, err, "Unable to set Status on NodeSet", &nodeSet)
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	if haveError {
+		return ctrl.Result{}, err
+	}
+
+	if shouldRequeue {
+		logger.Info("Not all NodeSets done for OpenStackDeployment")
+		return ctrl.Result{}, nil
```

Comment:
isn't it missing `RequeueAfter: time.Second` ?

Reply:
It's not needed since the Deployment owns the AnsibleEE's. As the AnsibleEE's change state, a reconcile for the Deployment will be automatically triggered. If we use RequeueAfter, we'll be looping needlessly.

That being said, I suppose the variable "shouldRequeue" is a misnomer, since we don't explicitly force a requeue. I could change that name to "shouldReturn" or "stillDeploying", etc.